boolean/boolean.py

```diff
@@ -1032,55 +1094,61 @@ def start_operation(ast, operation):
                 ast[0].append(ast[1](*ast[2:], eval=eval))
                 ast = ast[0]
 
-    expr = expr.replace(" ", "")
-    length = len(expr)
+    if isinstance(expr, str):
+        tokenized = tokenizer(expr)
+    else:
+        try:
+            tokenized = list(expr)
+        except:
+            raise TypeError("Argument must be string or an iterable of tokens but it is %s." % type(expr))
+
     ast = [None, None]
-    i = 0
-    while i < length:
-        char = expr[i]
-        if char == "1":
-            ast.append(TRUE)
-        elif char == "0":
-            ast.append(FALSE)
-        elif char.isalpha():
-            j = 1
-            while i + j < length and expr[i + j].isalnum():
-                j += 1
-            ast.append(Symbol(expr[i:i + j]))
-            i += j - 1
-        elif char == "(":
-            ast = [ast, "("]
-        elif char == ")":
-            while True:
-                if ast[0] is None:
-                    raise TypeError("Bad closing bracket at position %s." % i)
-                if ast[1] is "(":
-                    ast[0].append(ast[2])
-                    ast = ast[0]
-                    break
-                ast[0].append(ast[1](*ast[2:], eval=eval))
-                ast = ast[0]
-        elif char == "~":
-            ast = [ast, NOT]
-        elif char == "*":
-            ast = start_operation(ast, AND)
-        elif char == "+":
-            ast = start_operation(ast, OR)
+    for is_symbol, tok, srow, scol in tokenized:
+        if is_symbol:
+            if tok == "1":
```

Comment:
Maybe this should be `if tok in TRUES` (so it can maybe receive `true` or `True` as well as `1`) ...same for `FALSES`